Adding Lead and LeadBismuth Fluid properties

CalfSteven opened the pull request:
Used the known equations from "Handbook on Lead-bismuth Eutectic Alloy and Lead Properties "(2015). Added the known formulations for Rho, mu, cp, k and h. While adding other formulation that are needed.

CalfSteven committed:
adc Lead and LeadBismuth to Fluidproperties

fix up header

ref #21531

Fixed clang format for Lead and Lead Bismuth Fluid properties + headers

CalfSteven committed:
Added doc files, and bulk modulus and speed of sound formulation to source file and header

Added doc files, new citation to bib_fluid_properties, and bulk modulus and speed of sound formulation to source file and header

removed two cout statments that were left over from testing

gave led and led bismuth log descriptions

wrote descriptions in the header for speed of sound, corrected syntac in the bib and source miles, also wrote more info in the md file

wrote descriptions in the header for speed of sound, corrected syntac in the bib and source miles, also wrote more info in the md file
ref #21531

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1109758/) on 8f6f52b wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>d655d3</th>
      <th colspan=4>#21503 8f6f52</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>83.54%</td>
      <td>83.54%</td>
      <td>-0.01%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>77882</td>
      <td>77877</td>
      <td>-5</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15340</td>
      <td>15345</td>
      <td>+5</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21503/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21503/coverage/framework)


## Modules coverage

### Fluid properties

<table>
  <thead>
    <tr>
      <th></th>
      <th>d655d3</th>
      <th colspan=4>#21503 8f6f52</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.43%</td>
      <td>81.31%</td>
      <td>-4.12%</td>
      <td>2.35%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>5565</td>
      <td>5573</td>
      <td>+8</td>
      <td>8</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>949</td>
      <td>1281</td>
      <td>+332</td>
      <td>332</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21503/coverage_diff/fluid_properties)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21503/coverage/fluid_properties)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/21503/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/21503/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/21503/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/21503/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/21503/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/21503/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/21503/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/21503/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/21503/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/21503/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/21503/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/21503/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/21503/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/21503/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/21503/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/21503/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/21503/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/21503/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/21503/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/21503/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/21503/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/21503/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/21503/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/21503/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/21503/coverage/xfem)

</details>

## Warnings

- `fluid_properties` new line coverage rate 2.35% is less than the suggested 90.0%



This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1109758/) on 8f6f52b wanted to post the following:


The following coverage requirement(s) failed:
- `fluid_properties` coverage rate 81.31% is less than the required 84.0%

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1110612/) on ad39a81 wanted to post the following:


The following coverage requirement(s) failed:
- `fluid_properties` coverage rate 81.32% is less than the required 84.0%

GiudGiud committed:
Fix headers:
- remove inherited docstrings
- fix docstrings
- add missing routine for p_rho
- remove virtual where not needed
ref #21531

GiudGiud committed:
Fix leadBi fluid properties:
- remove approximarion of integration
- use handbook formula for c
- fix error in coefficient for k
ref #21531

GiudGiud committed:
Fix mistakes in lead fluid properties:
- wrong sign on derivative
- approximate coefficient during integration of cp
- use handbook expression for computing sound speed
ref #21531

GiudGiud committed:
Fix specific isochoric heat capacity following equiation 2.46 of NEA Lead handbook
ref #21531

Fix mistakes in derivatives of p_v_e and cp/cv_v_e

GiudGiud commented:
Should be mostly correct now.
This needs extensive unit testing to make sure it's 100% correct.

GiudGiud committed:
Upgrade documentation for lead and leadbismuth fluid properties
ref #21531

GiudGiud committed:
Add unit test for lead fluid properties

GiudGiud committed:
Add unit test for lead bismuth fluid properties
refs #21531

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1291938/) on c976eed wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21503/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21503/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0cdc064aa7d73392645a1a4f995a9a08497cd5a7`

GiudGiud commented:
@joshuahansel should be good for review if you have time

joshuahansel commented on modules/fluid_properties/doc/content/source/userobjects/LeadBismuthFluidProperties.md:
```diff
@@ -0,0 +1,44 @@
+| Density $(\rho, kg/m^3)$       | $11065 - 1.2793T$ |  2.30  |
+| Viscosity $(\mu, Pa \times s)$ | $4.94 \times 10^{-4} \times e^{\frac{754.1}{temperature}}$ |  2.83  |
+| Specific enthalpy (h, J)       | $164.8(T - T_{mo}) - 1.97 \times 10^{-2}(T^2 - T_{mo}^2) + 4.167 \times 10^6(T^3 - T_{mo}^3) + 4.56 \times 10^5(\frac{1}{T} -\frac{1}{T_{mo}})$ |  2.55  |
+| Thermal Conductivity $(k, \frac {W}{m * K})$ | $3.284 + 1.617 \times 10^{-2}T - 2.305 \times 10^{-6}T^2$ |  2.91  |
```
W/(m-K)

joshuahansel commented on modules/fluid_properties/doc/content/source/userobjects/LeadBismuthFluidProperties.md:
```diff
@@ -0,0 +1,44 @@
+| Viscosity $(\mu, Pa \times s)$ | $4.94 \times 10^{-4} \times e^{\frac{754.1}{temperature}}$ |  2.83  |
+| Specific enthalpy (h, J)       | $164.8(T - T_{mo}) - 1.97 \times 10^{-2}(T^2 - T_{mo}^2) + 4.167 \times 10^6(T^3 - T_{mo}^3) + 4.56 \times 10^5(\frac{1}{T} -\frac{1}{T_{mo}})$ |  2.55  |
+| Thermal Conductivity $(k, \frac {W}{m * K})$ | $3.284 + 1.617 \times 10^{-2}T - 2.305 \times 10^{-6}T^2$ |  2.91  |
+| Isobaric Specific Heat $(cp, \frac{J}{kg * K})$ | $164.8 - 3.94 * 10^{-2}T + 1.25 \times 10^{-5}T^2 - \frac{4.56 * 10^5}{T^2}$ |  Table 2.19.3  |
```
- $c_p$
- J/(kg-K)
- Don't use "*". Use either "\cdot" or "\times" and be consistent everywhere. Also be consistent with whether you put a multiplication symbol between the coefficient and the variable that follows, e.g, "10^{-5} T^2". I recommend not putting one there. You can add a space: "\,".

joshuahansel commented on modules/fluid_properties/doc/content/source/userobjects/LeadBismuthFluidProperties.md:
```diff
@@ -0,0 +1,44 @@
+
+| Properties     | Range/ Uncertainties|
+| :------------- | :------------- |
+| Density        | .8% |
```
0.8

joshuahansel commented on modules/fluid_properties/doc/content/source/userobjects/LeadFluidProperties.md:
```diff
@@ -0,0 +1,44 @@
+# LeadFluidProperties
```
Apply all of the comments from `LeadBismuthFluidProperties.md` here

joshuahansel commented on modules/fluid_properties/doc/content/source/userobjects/LeadFluidProperties.md:
```diff
@@ -0,0 +1,44 @@
+| Viscosity $(\mu, Pa \times s)$ | $4.55 \times 10^{-4} \times e^{\frac{1069}{temperature}}$ |  2.81  |
+| Specific enthalpy (h, J)       | $176.2(T - T_{mo}) - 2.4615 \times 10^{-2}(T^2 - T_{mo}^2) + 5.147 \times 10^6(T^3 - T_{mo}^3) + 1.524 \times 10^6(\frac{1}{T} -\frac{1}{T_{mo}})$ |  2.53  |
+| Thermal Conductivity $(k, \frac {W}{m * K})$ | $9.2 + 0.011T$ |  2.89  |
+| Isobaric Specific Heat $(cp, \frac{J}{kg * K})$ | $176.2 - 4.923* 10^{-2} \times T + 1.544 \times 10^-5 \times T^2 - \frac{1.524 * 10^6}{T^2}$ |  Table 2.19.1  |
```
10^{-5}

joshuahansel commented on modules/fluid_properties/include/userobjects/LeadBismuthFluidProperties.h:
```diff
@@ -0,0 +1,109 @@
+   *
+   * @param[in] p     pressure (Pa)
+   * @param[in] rho   (kg/m$^3$)
+   * @return[out] temperature (K)
```
@return[out] -> @return.

joshuahansel commented on modules/fluid_properties/include/userobjects/LeadFluidProperties.h:
```diff
@@ -0,0 +1,114 @@
+   *
+   * @param[in] p     pressure (Pa)
+   * @param[in] rho   density (kg/m$^3$)
+   * @return[out] temperature (K)
```
same here

joshuahansel commented on modules/fluid_properties/src/userobjects/LeadBismuthFluidProperties.C:
```diff
@@ -0,0 +1,386 @@
+LeadBismuthFluidProperties::validParams()
+{
+  InputParameters params = SinglePhaseFluidProperties::validParams();
+  params.addParam<Real>("T_mo", 398, "Melting Point of Lead Bismuth (K)");
```
Delete "of Lead Bismuth"

joshuahansel commented on modules/fluid_properties/src/userobjects/LeadBismuthFluidProperties.C:
```diff
@@ -0,0 +1,386 @@
+Real
+LeadBismuthFluidProperties::molarMass() const
+{
+  return 233.99;
```
This needs to be in SI, kg/mol

joshuahansel commented on modules/fluid_properties/src/userobjects/LeadFluidProperties.C:
```diff
@@ -0,0 +1,390 @@
+Real
+LeadFluidProperties::molarMass() const
+{
+  return 207.2;
```
This needs to be in SI, kg/mol

joshuahansel commented on modules/fluid_properties/src/userobjects/LeadBismuthFluidProperties.C:
```diff
@@ -0,0 +1,386 @@
+}
+
+void
+LeadBismuthFluidProperties::rho_from_p_T(const DualReal & pressure,
```
Use "p" and "T" for pressure and temperature everywhere in this file (do same for Lead)

joshuahansel commented on modules/fluid_properties/src/userobjects/LeadFluidProperties.C:
```diff
@@ -0,0 +1,390 @@
+  // Real e_p_plus = e_from_p_rho(p * (1 + epsilon), rho);
+  // Real e_rho_plus = e_from_p_rho(p, rho * (1 + epsilon));
+  // de_dp = (e_p_plus - e) / epsilon / p;
+  // de_drho = (e_rho_plus - e) / epsilon / rho;
```
Delete these comments

joshuahansel commented on modules/fluid_properties/src/userobjects/LeadFluidProperties.C:
```diff
@@ -0,0 +1,390 @@
+LeadFluidProperties::T_from_v_e(Real v, Real e, Real & T, Real & dT_dv, Real & dT_de) const
+{
+  T = T_from_v_e(v, e);
+  // Returning straight derivatives since no pressure dependence
```
Delete comment

joshuahansel commented on modules/fluid_properties/src/userobjects/LeadFluidProperties.C:
```diff
@@ -0,0 +1,390 @@
+LeadFluidProperties::validParams()
+{
+  InputParameters params = SinglePhaseFluidProperties::validParams();
+  params.addParam<Real>("T_mo", 600.6, "Melting Point of Lead (K)");
```
Delete "of Lead"

joshuahansel commented on modules/fluid_properties/unit/src/LeadBismuthFluidPropertiesTest.C:
```diff
@@ -0,0 +1,128 @@
+  const std::vector<Real> pressures = {1e5, 1e6, 5e6};
+  const std::vector<Real> temperatures = {400, 500, 700};
+
+  // Solutions : obtained from running the property object itself
+  // Verified visually against the reference
```
I'm confused. You got these values from the object you're testing? I know you haven't used full precision in these values, but your tolerances seem pretty loose. `REL_TOL_EXTERNAL_VALUE` is 1e-3. It's not meant for comparing to values from your own code. That's what `REL_TOL_SAVED_VALUE` (1e-12) is recommended for (requires outputting your values in all or most of their precision).

joshuahansel commented on modules/fluid_properties/unit/src/LeadBismuthFluidPropertiesTest.C:
```diff
@@ -0,0 +1,128 @@
+    const Real rho = _fp->rho_from_p_T(p, T);
+    const Real v = 1. / rho;
+    const Real e = _fp->e_from_p_T(p, T);
+
```
Need tests for

- p_from_v_e
- v_from_p_T

I'm guessing same for Lead

joshuahansel commented on modules/fluid_properties/unit/src/LeadBismuthFluidPropertiesTest.C:
```diff
@@ -0,0 +1,128 @@
+  const Real rho = _fp->rho_from_p_T(p, T);
+  const Real v = 1. / rho;
+  const Real e = _fp->e_from_p_T(p, T);
+
```
Need test for `c_from_v_e` (same for Lead)

GiudGiud commented on modules/fluid_properties/doc/content/source/userobjects/LeadBismuthFluidProperties.md:
```diff
@@ -0,0 +1,44 @@
+!table caption=Table of properties and references to the equations in [!citep](Fazio).
+| Properties                     | Equations| Equation # |
+| :----------------------------- | :------------- | :------------- |
+| Melting point $(T_{mo}, K)$    | 398 |  2.1  |
```
mo is probably molten

GiudGiud replied:
did (K) instead of [K] as moosedocs [k] is a link

joshuahansel replied:
For future reference, square brackets can be escaped with backslashes. I actually wrote it this way, but Github decided to be "smart" and render it with Markdown so that you can't see it 😛

GiudGiud committed:
Address Josh's review:
- pressure -> p; temperature -> T in variable names
- clean up table, use math style only as appropriate, homogenize equation and unit format
- use saved values to a higher precision for unit testing
- use correct unit for molar mass: kg/mol instead of g/mol

GiudGiud commented:
Thanks for the review!
